rclcpp/src/rclcpp/qos_event.cpp

```diff
@@ -0,0 +1,59 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/qos_event.hpp"
+
+
+namespace rclcpp
+{
+
+QOSEventHandlerBase::~QOSEventHandlerBase()
+{
+  if (rcl_event_fini(&event_handle_) != RCL_RET_OK) {
+    RCUTILS_LOG_ERROR_NAMED(
+      "rclcpp",
+      "Error in destruction of rcl event handle: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+  }
+}
+
+/// Get the number of ready events
+size_t
+QOSEventHandlerBase::get_number_of_ready_events()
+{
+  return 1;
+}
+
+/// Add the Waitable to a wait set.
+bool
+QOSEventHandlerBase::add_to_wait_set(rcl_wait_set_t * wait_set)
+{
+  if (rcl_wait_set_add_event(wait_set, &event_handle_, &wait_set_event_index_) != RCL_RET_OK) {
+    RCUTILS_LOG_ERROR_NAMED(
+      "rclcpp",
+      "Couldn't add event to wait set: %s", rcl_get_error_string().str);
```

Comment:
Any reason to not throw here?

Reply:
The only other examples of implementing `Waitable::add_to_wait_set()` are in `ClientBase` and `ServerBase` of `rclcpp_action`, and they do not throw.

The other `rcl_wait_set_add_*()` related functions found alongside `QOSEventHandlerBase::add_to_wait_set()` in `AllocatorMemoryStrategy::add_handles_to_wait_set()` don't seem to throw either.

Reply:
@sloretz or @jacobperron any opinion on this? I feel like this is neither something you can ignore, nor something that's ever expected, so I don't know why we wouldn't throw here. Can you guys imagine why?

Reply:
I'm not sure. The equivalent API in `rclpy` will throw: 

* https://github.com/ros2/rclpy/blob/616e853455bfd236e61ef09f92b4c5afe15af389/rclpy/rclpy/action/client.py#L332-L334
* https://github.com/ros2/rclpy/blob/616e853455bfd236e61ef09f92b4c5afe15af389/rclpy/src/rclpy/_rclpy_action.c#L172-L175

Reply:
Can you please update this so that it throws? I would also appreciate it if you could open an issue linking to the other places where you think that the same thing is happening, so we can fix that up later.

Reply:
I have added the exception throwing to `QOSEventHandlerBase::add_to_wait_set()`.

I have also opened the relevant issue at https://github.com/ros2/rclcpp/issues/706.

Reply:
Thanks!